dll/win32/shell32/shellrecyclebin/recyclebin_v5.cpp

```diff
@@ -672,31 +645,80 @@ RecycleBin5_Create(
     return hr;
 }
 
+RecycleBin5::RecycleBin5(
+    LPCWSTR VolumePath,
+    LPCWSTR RecycleBinDirectory,
+    LPWSTR StringSid,
+    PTOKEN_USER TokenUserInfo,
+    LPWSTR pszFolder)
+    : m_ref(1)
+    , m_hInfo(NULL)
+    , m_hInfoMapped(NULL)
+    , m_EnumeratorCount(0)
+    , m_VolumePath(NULL)
+    , m_Folder(pszFolder)
+    , m_hr(S_OK)
+{
+    INT len;
+    LPWSTR p;
+    HRESULT hr;
+
+    if (StringSid)
+        len = swprintf(m_Folder, L"%s%s\\%s", VolumePath, RecycleBinDirectory, StringSid);
+    else
+        len = swprintf(m_Folder, L"%s%s", VolumePath, RecycleBinDirectory);
+
+    p = &m_Folder[len];
+    wcscpy(p, L"\\" RECYCLE_BIN_FILE_NAME);
+
+    m_hInfo = CreateFileW(m_Folder, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
+    if (m_hInfo == INVALID_HANDLE_VALUE &&
+        (GetLastError() == ERROR_PATH_NOT_FOUND || GetLastError() == ERROR_FILE_NOT_FOUND))
+    {
+        *p = UNICODE_NULL;
+        hr = RecycleBin5_Create(m_Folder, TokenUserInfo ? TokenUserInfo->User.Sid : NULL);
+        *p = L'\\';
+        if (!SUCCEEDED(hr))
+        {
+            m_hr = hr;
+            return;
+        }
+        m_hInfo = CreateFileW(m_Folder, GENERIC_READ | GENERIC_WRITE, 0, NULL, OPEN_EXISTING, 0, NULL);
+    }
+    if (m_hInfo == INVALID_HANDLE_VALUE)
+    {
+        m_hr = HRESULT_FROM_WIN32(GetLastError());
+        return;
+    }
+    m_hInfoMapped = CreateFileMappingW(m_hInfo, NULL, PAGE_READWRITE | SEC_COMMIT, 0, 0, NULL);
+    if (!m_hInfoMapped)
+    {
+        m_hr = HRESULT_FROM_WIN32(GetLastError());
+        return;
+    }
+    *p = UNICODE_NULL;
+    m_VolumePath = p + 1;
+    wcscpy(m_VolumePath, VolumePath);
+}
+
+EXTERN_C
 HRESULT RecycleBin5_Constructor(IN LPCWSTR VolumePath, OUT IUnknown **ppUnknown)
```

Comment:
Similarly to the RecycleBin5Enum PR, I wonder why not more of the code contained here isn't in the actual C++ constructor of the RecycleBin5 class.

Reply:
Used `Init()`.